tests/testcfg.rs

```diff
@@ -0,0 +1,107 @@
+use tempfile::tempdir_in;
+
+pub enum FuelupState {
+    Empty,
+    LatestToolchainInstalled,
+}
+
+pub struct TestCfg {
+    pub fuelup_path: PathBuf,
+    pub root: PathBuf,
+    pub home: PathBuf,
+}
+
+#[derive(Debug)]
+pub struct TestOutput {
+    pub stdout: String,
+    pub stderr: String,
+}
+
+impl TestCfg {
+    pub fn new(fuelup_path: PathBuf, root: PathBuf, home: PathBuf) -> Self {
+        Self {
+            fuelup_path,
+            root,
+            home,
+        }
+    }
+
+    pub fn toolchains_dir(&self) -> PathBuf {
+        self.home.join(".fuelup").join("toolchains")
+    }
+
+    pub fn toolchain_bin_dir(&self, toolchain: &str) -> PathBuf {
+        self.home
+            .join(".fuelup")
+            .join("toolchains")
+            .join(toolchain)
+            .join("bin")
+    }
+
+    pub fn fuelup(&mut self, args: &[&str]) -> TestOutput {
+        let output = Command::new(&self.fuelup_path)
+            .args(args)
+            .env("HOME", &self.home)
+            .output()
+            .expect("Failed to execute command");
+        let stdout = String::from_utf8(output.stdout).unwrap();
+        let stderr = String::from_utf8(output.stderr).unwrap();
+        TestOutput { stdout, stderr }
+    }
+}
+
+pub fn setup(state: FuelupState, f: &dyn Fn(&mut TestCfg)) -> Result<()> {
+    let root = env::current_exe()
+        .unwrap()
+        .parent()
+        .expect("fuelup's directory")
+        .to_path_buf();
+
+    let testdir = tempdir_in(&root).unwrap();
+    let tmp_home = testdir.path();
+
+    let tmp_fuelup_root_path = tmp_home.join(".fuelup");
+    let tmp_fuelup_bin_dir_path = tmp_home.join(".fuelup").join("bin");
+    fs::create_dir(&tmp_fuelup_root_path).unwrap();
+    fs::create_dir(&tmp_fuelup_bin_dir_path).unwrap();
+    fs::create_dir(&tmp_fuelup_root_path.join("toolchains")).unwrap();
+    fs::copy(
+        root.parent().unwrap().join("fuelup"),
+        &tmp_fuelup_bin_dir_path.join("fuelup"),
+    )?;
+
+    match state {
+        FuelupState::Empty => {}
+        FuelupState::LatestToolchainInstalled => {
+            let bin_dir = tmp_fuelup_root_path
+                .join("toolchains")
+                .join("latest-x86_64-apple-darwin")
+                .join("bin");
+            fs::create_dir_all(&bin_dir)
+                .expect("Failed to create temporary latest toolchain bin dir");
+
+            fs::File::create(&bin_dir.join("forc"))?;
+            fs::File::create(&bin_dir.join("forc-fmt"))?;
+            fs::File::create(&bin_dir.join("forc-lsp"))?;
+            fs::File::create(&bin_dir.join("forc-explore"))?;
+            fs::File::create(&bin_dir.join("fuel-core"))?;
+            fs::copy(
+                &env::current_dir()
+                    .unwrap()
+                    .join("tests/settings-example.toml"),
+                &tmp_fuelup_root_path.join("settings.toml"),
+            )
+            .expect("Failed to copy settings");
+        }
```

Comment:
Is the idea behind this `LatestToolchainInstalled` branch to emulate having `fuelup` in a state where `latest` is already installed? Could you elaborate a bit on why we need this, and why we manually create stub dirs rather than doing `fuelup toolchain install latest` to have it auto-populate with the binaries etc too?

Reply:
Yes, the main idea is to emulate `fuelup` in a state where `latest` is already installed. I think it'd be ideal to have as little tests that rely on the network as possible, and hence that is also why I combined the `install` and `check` tests into the same test.

In the case of certain commands like `fuelup component remove <component>`, downloading seems like an unnecessary step, since all we want to do is to remove something called `forc` for example and then ensure that it isn't there, without deleting other components. In this case we can just have a file stub called `forc` and we wouldn't need to download the actual `forc`. Same for when uninstalling toolchains - we just want to make sure that a toolchain directory is no longer there.

Reply:
An alternative is to download the binaries once at the start of `cargo test` and reuse them whereever needed by copying them into the required toolchain

Reply:
> In the case of certain commands like fuelup component remove <component>, downloading seems like an unnecessary step, since all we want to do is to remove something called forc for example and then ensure that it isn't there, without deleting other components. In this case we can just have a file stub called forc and we wouldn't need to download the actual forc. Same for when uninstalling toolchains - we just want to make sure that a toolchain directory is no longer there.

Yeah I can empathise with wanting to reduce network I/O for tests. My concern is that we end up testing for a case that can never occur anyway, and it might end up being more of a pain to synchronise this set of stubs with the actual behaviour of fuelup as we add different plugins and so-on :thinking: 

> An alternative is to download the binaries once at the start of cargo test and reuse them whereever needed by copying them into the required toolchain

Ahh this would be nice, though I'm not sure how easy it would be to do this in a manner that still allows for separate tests? I guess we could have a single `#[test] fn integration() { ... }` or something like that which downloaded first and then ran the other tests as functions in order one at a time? It would be a shame to not have an individual `#[test]` for each command, but perhaps worth it to remove the need for all the extra networking and to avoid the need to fake a fuelup installation with stubs?

Reply:
>An alternative is to download the binaries once at the start of cargo test and reuse them whereever needed by copying them into the required toolchain

Perhaps this isn't possible after all and I spoke too soon? Not sure if setting everything up before all tests would be possible. Seems like it is still an open RFC (https://github.com/rust-lang/rfcs/issues/1664).

> I guess we could have a single #[test] fn integration() { ... } or something like that which downloaded first and then ran the other tests as functions in order one at a time? It would be a shame to not have an individual #[test] for each command, but perhaps worth it to remove the need for all the extra networking and to avoid the need to fake a fuelup installation with stubs?

Hmmm.. I guess this would work as a workaround but it just seems like a tradeoff between the code complexity vs. faking installation with file stubs... in which case it feels like stubs are a 'good-enough' solution IMO since in most cases (for now) we would only want to test the file/directory based manipulation around managing toolchains/components... unless we eventually want to test advanced features. For example, features that would require us to actually execute the real binaries and check the stdout/stderr of what `forc` produces, for example. But it seems to me that testing fuelup's role in such an example would just be testing if the correct executable is called.

Reply:
OK let's see how these tests go for now. If it turns out these stubs end up causing too much churn we can revisit this in the future.

Reply:
I think it's a result of the error not piping to stderr for some weird reason though it works on my local tests. My guess that it somehow was printed to ardour instead in the CI. I haven't figured out why though...